// dooboo-ui/Slider/Thumb.js
import { Animated } from 'react-native';
import React from 'react';
import styled from '@emotion/native';
const ThumbPositioner = styled.View `
  position: absolute;
  left: ${({ percent }) => `${percent}%`};
`;
const StyledThumb = styled.View `
  display: flex;
  justify-content: center;
  align-items: center;
  width: 12px;
  height: 12px;
  background-color: #0b21e8;
  border-radius: 6px;
  transform: translate(-6px);
`;
const Thumb = ({ testID, percent, size = 12, scaleValue = new Animated.Value(0.01), opacityValue = new Animated.Value(0.12), customThumb, style, }) => {
    const rippleSize = (size - 2) * 2 + 48 / size;
    return (<ThumbPositioner testID="thumb-positioner-test-id" percent={percent}>
      {customThumb || (<StyledThumb testID={testID} style={style}>
          <Animated.View testID="thumb-animated" style={{
                width: rippleSize,
                height: rippleSize,
                borderRadius: 100,
                transform: [{ scale: scaleValue }],
                opacity: opacityValue,
                backgroundColor: '#0b21e8',
            }}/>
        </StyledThumb>)}
    </ThumbPositioner>);
};
export default Thumb;
