// dooboo-ui/Slider/utils.js
export const roundNearest = (number, digit) => {
    const rest = number % digit;
    const halfDigit = digit / 2;
    const base = number - rest;
    if (rest >= halfDigit)
        return base + digit;
    return base;
};
export const getPercentByValue = (value, maxValue, minValue) => {
    return (value / (maxValue - minValue)) * 100;
};
export const getValueByPercent = (percent, maxValue, minValue) => {
    return ((maxValue - minValue) * percent) / 100;
};
export const getStepPercent = ({ minValue, maxValue, step, }) => {
    return (100 / (maxValue - minValue)) * step;
};
export const getPercentByPositionX = ({ positionX, sliderWidth, stepPercent, }) => {
    const percent = (positionX / sliderWidth) * 100;
    if (percent <= 0)
        return 0;
    if (percent >= 100)
        return 100;
    return roundNearest(percent, stepPercent);
};
export const getNearestPercentByValue = ({ value, minValue, maxValue, step, }) => {
    const stepPercent = getStepPercent({
        minValue,
        maxValue,
        step,
    });
    const percent = getPercentByValue(value, maxValue, minValue);
    if (percent <= 0)
        return 0;
    if (percent >= 100)
        return 100;
    return roundNearest(getPercentByValue(value, maxValue, minValue), stepPercent);
};
export const getStepValueByPercent = ({ percent, stepPercent, step, }) => {
    return (percent / stepPercent) * step;
};
