// dooboo-ui/SwitchToggle.js
import { Animated, StyleSheet, Text, TouchableOpacity, View, } from 'react-native';
import React, { useEffect, useRef, useState } from 'react';
const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
    },
});
function SwitchToggle(props) {
    const [animXValue] = useState(new Animated.Value(props.switchOn ? 1 : 0));
    const getStart = () => {
        // prettier-ignore
        return props.type === undefined
            ? 0
            : props.type === 0
                ? 0
                : props.containerStyle && props.containerStyle.padding
                    ? props.containerStyle.padding * 2
                    : {};
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
    const runAnimation = () => {
        const animValue = {
            fromValue: props.switchOn ? 0 : 1,
            toValue: props.switchOn ? 1 : 0,
            duration: props.duration,
            useNativeDriver: false,
        };
        Animated.timing(animXValue, animValue).start();
    };
    const endPos = props.containerStyle && props.circleStyle
        ? props.containerStyle.width -
            (props.circleStyle.width +
                (props.containerStyle.padding || 0) * 2)
        : 0;
    const circlePosXEnd = props.RTL ? -endPos : endPos;
    const [circlePosXStart] = useState(getStart());
    const prevSwitchOnRef = useRef();
    const prevSwitchOn = !!prevSwitchOnRef.current;
    useEffect(() => {
        prevSwitchOnRef.current = props.switchOn;
        if (prevSwitchOn !== props.switchOn)
            runAnimation();
    }, [prevSwitchOn, props.switchOn, runAnimation]);
    const generateRightText = () => {
        return (<Animated.View style={props.rightContainerStyle}>
        <Text style={props.textRightStyle}>{props.backTextRight}</Text>
      </Animated.View>);
    };
    const generateLeftText = () => {
        return (<Animated.View style={props.leftContainerStyle}>
        <Text style={props.textLeftStyle}>{props.backTextLeft}</Text>
      </Animated.View>);
    };
    const generateLeftIcon = () => {
        return (<View style={{ position: 'absolute', left: 5 }}>
        {props.backgroundImageOn}
      </View>);
    };
    const generateRightIcon = () => {
        return (<View style={{ position: 'absolute', right: 5 }}>
        {props.backgroundImageOff}
      </View>);
    };
    return (<TouchableOpacity testID={props.testID} onPress={props.onPress} activeOpacity={0.5}>
      <Animated.View style={[
            styles.container,
            props.containerStyle,
            {
                backgroundColor: animXValue.interpolate({
                    inputRange: [0, 1],
                    outputRange: [
                        props.backgroundColorOff,
                        props.backgroundColorOn,
                    ],
                }),
            },
        ]}>
        {generateLeftText()}
        {props.switchOn && generateLeftIcon()}
        <Animated.View style={[
            props.circleStyle,
            {
                backgroundColor: animXValue.interpolate({
                    inputRange: [0, 1],
                    outputRange: [
                        props.circleColorOff,
                        props.circleColorOn,
                    ],
                }),
            },
            {
                transform: [
                    {
                        translateX: animXValue.interpolate({
                            inputRange: [0, 1],
                            outputRange: [
                                circlePosXStart,
                                circlePosXEnd,
                            ],
                        }),
                    },
                ],
            },
            props.buttonStyle,
        ]}>
          <Animated.View style={props.buttonContainerStyle}>
            <Text style={props.buttonTextStyle}>{props.buttonText}</Text>
          </Animated.View>
        </Animated.View>
        {generateRightText()}
        {!props.switchOn && generateRightIcon()}
      </Animated.View>
    </TouchableOpacity>);
}
export { SwitchToggle };
